### ISSUE TITLE
LSTM autoencoder for audio - input shape? [Sequence to sequence]

### ISSUE THREAD

**Issue by dhrushilbadani** [fchollet/keras#4523]
So, I'm trying to learn fixed vector representations for segments of about 200 songs (~ 3-5 minutes per song) and wanted to use an LSTM-based Sequence-to-sequence Autoencoder for it.

I'm preprocessing the audio (using [librosa](librosa.github.io)) as follows:

1. I'm first just getting a raw audio signal time series of shape around (1500000,) - (2500000,) per song.
2. I'm then slicing each raw time series into segments and getting a lower-level mel spectrogram matrix of shape (512, 3000) - (512, 6000) per song. Each of these (512,) vectors can be referred to as 'mini-songs' as they represent parts of the song.
3. I vertically stack all these mini-songs of all the songs together to create the training data (let's call this **X**). X turns out to be (512, 600000) in size, where the first dimension (512) is the window size and the second dimension (600000) is the total number of 'mini-songs' in the dataset.
4. Which is to say, there are about 600000 mini-songs in X - each column in X represents a mini-song of length (512,).
5. Each of these (512,) mini-song vectors should be encoded into a (50,) vector per mini-song i.e. we will have 600000 (50,) vectors at the end of the process.
6. In more standard terminology, I have 600000 training samples each of length 512. [Think of this as being similar to an image dataset - 600000 images, each of length 784, where the images are of resolution 32x32. Except in my case I want to treat the 512-length samples as sequences that have temporal properties.]

I read the example [here](https://blog.keras.io/building-autoencoders-in-keras.html) and was looking to extend that for my use case. I was wondering what the `timesteps` and `input_dim` parameters to the `Input` layer should be set to. 

I'm setting `timesteps = X.shape[0]` (i.e. 512 in this case) and `input_dim = X.shape[1]` (i.e 600000). Is this the correct way to go about it?

**Edit:** Added clarifications above.

**Comment by santi-pdp** [comment]
It looks like you are inverting the dimensions as supposed to be. You say your axis 0 are the features  (fixed dimension of 512), and the axis 1 varies depending on the sequence length (600000 - 1200000) and the number of songs you have, so you are interleaving the sequences in axis 1. 
The way to go with recurrent neural networks is building a 3D tensor, ass follows: you stack your different sequences in the axis 0 (samples dimension), you set your variable dimension in the axis 1 (sequence length), and then finally set you fixed-dimensional features in the axis 2 (features dimension). In your case, with the numbers you've given, you should obtain a matrix: X (200, seqlength, 512). In seqlength you can set the maximum you may find in your training set (6000) and pad your sequences such that real content appears after PAD content in the input, and before PAD content in the output. After this, `timesteps=X.shape[1] `and `input_dim=X.shape[2]`

**Comment by dhrushilbadani** [comment]
@santi-pdp Thank you for your prompt response! Just to clarify - there's really _no_ "variable dimension." Each song is being split into 'mini-songs' [represented by a shape (512,) vector of mel frequency values] via librosa. It is these mini song vectors that I want to convert to a fixed-length vector representation [say a shape (50,) vector per mini-song] using the autoencoder:
1. I have about 600000 'mini-songs'
2. Each of these mini-songs is represented by a low-level (512,) mel-frequency vector
3. The total training data matrix is thus (512, 600000).
4. Each of these (512,) mini-song vectors should be encoded into one (50,) vector per mini-song i.e. we will have 600000 (50,) vectors at the end of the process.
5. In more standard terminology, I have 600000 'mini-song samples' each of length '512'. [Think of this as being similar to an image dataset - 600000 images, each of length 784, were the images of resolution 32x32]

How then would you recommend I go about this?

**Comment by santi-pdp** [comment]
@dhrushilbadani If you have to process sequentially all those vectors, then your structure would be: X (600000, 512, 1), so you have to transpose and reshape your X matrix:
`X = X.T`
`X = X.reshape(X.shape + (1,))`
your input_dim is 1 (sample by sample from 512 dimensional vector) and your timestemps are 512

**Comment by dhrushilbadani** [comment]
@santi-pdp Thanks! I'm using the code provided [here](https://blog.keras.io/building-autoencoders-in-keras.html) by the developers of Keras to do something similar to the approach tried by [Dai and Le (2015)](https://papers.nips.cc/paper/5949-semi-supervised-sequence-learning.pdf) :

```
from keras.layers import Input, LSTM, RepeatVector
from keras.models import Model

# X is of shape (600000, 512, 1)
# I want to learn a (50,) representation of each sequence

latent_dim = 10
timesteps = 512
input_dim = 1

inputs = Input(shape=(timesteps, input_dim))
encoded = LSTM(latent_dim)(inputs)

decoded = RepeatVector(timesteps)(encoded)
decoded = LSTM(input_dim, return_sequences=True)(decoded)

sequence_autoencoder = Model(inputs, decoded)
encoder = Model(inputs, encoded)
```

What exactly is the RepeatVector layer doing here? And shouldn't the output of the `decoded` layer be of shape (timesteps,) instead of (input_dim,)?

**Comment by keunwoochoi** [comment]
I guess the output shape of `decoded` == `(None, timesteps, input_dim)`, no?

The `RepeatVector` repeats the input vector (which is `encoded` here) so that it can be a proper input of `LSTM(input_dim, return_sequences=True)`. I don't understand why the authors had to do like this though.

`model.summary()` is useful, try it!
```
>>> sequence_autoencoder.summary()
____________________________________________________________________________________________________
Layer (type)                     Output Shape          Param #     Connected to
====================================================================================================
input_1 (InputLayer)             (None, 512, 1)        0
____________________________________________________________________________________________________
lstm_1 (LSTM)                    (None, 50)            10400       input_1[0][0]
____________________________________________________________________________________________________
repeatvector_1 (RepeatVector)    (None, 512, 50)       0           lstm_1[0][0]
____________________________________________________________________________________________________
lstm_2 (LSTM)                    (None, 512, 1)        208         repeatvector_1[0][0]
====================================================================================================
Total params: 10608
____________________________________________________________________________________________________
```
